File: scripts/solar/util/archive/ArchiveUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 2,["8"] = 2,["9"] = 3,["10"] = 3,["11"] = 4,["12"] = 4,["13"] = 14,["14"] = 14,["15"] = 14,["17"] = 14,["18"] = 31,["19"] = 31,["20"] = 31,["22"] = 32,["23"] = 33,["24"] = 34,["26"] = 36,["27"] = 37,["28"] = 38,["30"] = 40,["31"] = 31,["32"] = 50,["33"] = 50,["34"] = 50,["36"] = 51,["37"] = 52,["38"] = 53,["39"] = 54,["40"] = 55,["42"] = 57,["43"] = 58,["44"] = 59,["46"] = 50,["47"] = 66,["48"] = 67,["49"] = 68,["50"] = 66,["51"] = 75,["52"] = 76,["53"] = 77,["54"] = 78,["57"] = 81,["58"] = 82,["59"] = 83,["61"] = 85,["62"] = 75,["63"] = 91,["65"] = 92,["66"] = 92,["67"] = 93,["68"] = 94,["69"] = 92,["72"] = 96,["73"] = 91,["74"] = 102,["75"] = 103,["76"] = 104,["77"] = 105,["78"] = 106,["80"] = 109,["81"] = 110,["82"] = 111,["83"] = 113,["84"] = 114,["85"] = 115,["87"] = 117,["88"] = 118,["89"] = 119,["90"] = 120,["91"] = 121,["95"] = 124,["96"] = 124,["97"] = 125,["98"] = 126,["99"] = 127,["101"] = 129,["102"] = 130,["103"] = 131,["104"] = 124,["107"] = 134,["108"] = 135,["109"] = 102,["110"] = 145,["111"] = 146,["112"] = 147,["114"] = 149,["116"] = 150,["117"] = 150,["118"] = 151,["119"] = 153,["120"] = 154,["122"] = 150,["125"] = 158,["126"] = 145,["127"] = 161,["128"] = 162,["129"] = 163,["130"] = 164,["132"] = 165,["133"] = 165,["134"] = 166,["135"] = 167,["136"] = 168,["137"] = 169,["141"] = 165,["144"] = 175,["145"] = 176,["149"] = 219,["150"] = 220,["151"] = 220,["152"] = 220,["153"] = 220,["154"] = 220,["155"] = 220,["156"] = 220,["157"] = 221,["160"] = 179,["161"] = 180,["163"] = 182,["164"] = 183,["165"] = 185,["166"] = 186,["167"] = 187,["168"] = 188,["169"] = 188,["170"] = 188,["171"] = 189,["172"] = 188,["173"] = 188,["175"] = 193,["176"] = 194,["177"] = 195,["178"] = 197,["179"] = 199,["180"] = 200,["181"] = 201,["183"] = 204,["184"] = 205,["185"] = 206,["187"] = 209,["188"] = 210,["189"] = 210,["190"] = 210,["191"] = 210,["192"] = 210,["193"] = 210,["194"] = 210,["195"] = 211,["196"] = 211,["197"] = 211,["198"] = 212,["199"] = 211,["200"] = 211,["203"] = 217,["209"] = 178,["212"] = 223,["213"] = 161,["214"] = 16,["215"] = 17,["216"] = 18,["217"] = 19,["218"] = 20,["219"] = 22,["220"] = 143,["221"] = 14});
local ____exports = {}
local ____CodecUtil = require("solar.util.math.CodecUtil")
local CodecUtil = ____CodecUtil.default
local ____BaseUtil = require("solar.util.BaseUtil")
local BaseUtil = ____BaseUtil.default
local ____player = require("solar.w3ts.handles.player")
local MapPlayer = ____player.MapPlayer
____exports.default = __TS__Class()
local ArchiveUtil = ____exports.default
ArchiveUtil.name = "ArchiveUtil"
function ArchiveUtil.prototype.____constructor(self)
end
function ArchiveUtil.get(self, whichPlayer, key, mapLevelLimit)
    if mapLevelLimit == nil then
        mapLevelLimit = 0
    end
    ____exports.default:_init0()
    if mapLevelLimit > 0 and DzAPI_Map_GetMapLevel(whichPlayer) < mapLevelLimit then
        return nil
    end
    local data = ____exports.default.playerDatas["P" .. tostring(GetPlayerId(whichPlayer))]
    if not data then
        return nil
    end
    return data[key]
end
function ArchiveUtil.set(self, whichPlayer, key, value, isSaveAll)
    if isSaveAll == nil then
        isSaveAll = true
    end
    ____exports.default:_init0()
    local data = ____exports.default.playerDatas["P" .. tostring(GetPlayerId(whichPlayer))]
    if not data then
        data = {}
        ____exports.default.playerDatas["P" .. tostring(GetPlayerId(whichPlayer))] = data
    end
    data[key] = value
    if isSaveAll then
        ____exports.default:saveAll(whichPlayer)
    end
end
function ArchiveUtil.getAllArchive(self, whichPlayer)
    ____exports.default:_init0()
    return ____exports.default.playerDatas["P" .. tostring(GetPlayerId(whichPlayer))]
end
function ArchiveUtil.printAllArchive(self, whichPlayer)
    ____exports.default:_init0()
    local data = ____exports.default.playerDatas["P" .. tostring(GetPlayerId(whichPlayer))]
    if not data then
        return
    end
    print(("========打印玩家" .. tostring(GetPlayerId(whichPlayer) + 1)) .. "的存档 开始========")
    for dataKey in pairs(data) do
        print(((dataKey .. "=[") .. tostring(data[dataKey])) .. "]")
    end
    print(("========打印玩家" .. tostring(GetPlayerId(whichPlayer) + 1)) .. "的存档 结束========")
end
function ArchiveUtil.clearAllArchive(self, whichPlayer)
    do
        local i = 0
        while i < ____exports.default.ArchiveKeyCountMax do
            local key = ____exports.default.baseArchiveKeyPrefix .. tostring(i)
            DzAPI_Map_FlushStoredMission(whichPlayer, key)
            i = i + 1
        end
    end
    print("已清理全部存档!")
end
function ArchiveUtil.saveAll(self, whichPlayer)
    local data = ____exports.default.playerDatas["P" .. tostring(GetPlayerId(whichPlayer))]
    if not data then
        data = {}
        ____exports.default.playerDatas["P" .. tostring(GetPlayerId(whichPlayer))] = data
    end
    data._SL = DzAPI_Map_GetMapLevel(whichPlayer)
    data._SH = StringHash(GetPlayerName(whichPlayer))
    data._SI = MapPlayer:fromHandle(whichPlayer).userId
    local dataJsonStr = JSON:stringify(data)
    if ____exports.default.encrypt then
        dataJsonStr = CodecUtil:saesEncode(dataJsonStr, gameName .. ____exports.default.salt)
    end
    local dataJsonStrLength = #dataJsonStr
    local keyLength = math.floor(dataJsonStrLength / 64) + 1
    log.debug((("玩家" .. tostring(GetPlayerId(whichPlayer))) .. "的存档栏位大小为:") .. tostring(keyLength))
    if keyLength > ____exports.default.ArchiveKeyCountMax then
        log.errorWithTraceBack((("存档栏位大小超过" .. tostring(____exports.default.ArchiveKeyCountMax)) .. "！") .. tostring(keyLength))
        return
    end
    do
        local i = 0
        while i < keyLength do
            local endIndex = (i + 1) * 64
            if endIndex > dataJsonStrLength then
                endIndex = dataJsonStrLength
            end
            local oneData = __TS__StringSubstring(dataJsonStr, i * 64, endIndex)
            local key = ____exports.default.baseArchiveKeyPrefix .. tostring(i)
            DzAPI_Map_SaveServerValue(whichPlayer, key, oneData)
            i = i + 1
        end
    end
    local key = ____exports.default.baseArchiveKeyPrefix .. tostring(keyLength)
    DzAPI_Map_FlushStoredMission(whichPlayer, key)
end
function ArchiveUtil._init0(self)
    if ____exports.default.isInit then
        return true
    end
    ____exports.default.isInit = true
    do
        local i = 0
        while i < bj_MAX_PLAYER_SLOTS do
            local tempPlayer = Player(i)
            if GetPlayerController(tempPlayer) == MAP_CONTROL_USER and GetPlayerSlotState(tempPlayer) == PLAYER_SLOT_STATE_PLAYING then
                ____exports.default.playerDatas["P" .. tostring(i)] = ____exports.default:_init0_onePlayer(tempPlayer)
            end
            i = i + 1
        end
    end
    return true
end
function ArchiveUtil._init0_onePlayer(self, tempPlayer)
    local data = {}
    local i = GetPlayerId(tempPlayer)
    local dataJsonStr = ""
    do
        local i = 0
        while i < ____exports.default.ArchiveKeyCountMax do
            local key = ____exports.default.baseArchiveKeyPrefix .. tostring(i)
            local serverValue = DzAPI_Map_GetServerValue(tempPlayer, key)
            if serverValue and #serverValue > 0 then
                dataJsonStr = dataJsonStr .. serverValue
            else
                break
            end
            i = i + 1
        end
    end
    if #dataJsonStr < 2 then
        return {}
    end
    do
        local function ____catch(e)
            log.errorWithTraceBack("存档Json格式解析出错!")
            DisplayTimedTextToPlayer(
                GetLocalPlayer(),
                0,
                0,
                60,
                "存档格式解析出错！"
            )
            print(("需要解析的Json字符串为[" .. dataJsonStr) .. "]")
        end
        local ____try, ____hasReturned, ____returnValue = pcall(function()
            if ____exports.default.encrypt then
                dataJsonStr = CodecUtil:saesDecode(dataJsonStr, gameName .. ____exports.default.salt)
            end
            data = JSON:parse(dataJsonStr)
            log.debug((("玩家" .. tostring(i + 1)) .. ":的存档数据为:") .. dataJsonStr)
            if DzAPI_Map_GetMapLevel(tempPlayer) < data._SL then
                print("存档验证失败!默认使用空存档!")
                data = {}
                BaseUtil.runLater(
                    1,
                    function()
                        CustomDefeatBJ(tempPlayer, "警告:存档验证失败C1！")
                    end
                )
            end
            local shTemp = StringHash(GetPlayerName(tempPlayer))
            local userId = MapPlayer:fromHandle(tempPlayer).userId
            if userId == -1 then
                local info = "提示:请不要再地图初始时获取存档！应当延迟执行存档逻辑!"
                print(info)
                print("请等待内置userId(用户id)同步完成！(为了解决玩家名更改后靠userId判断是否是自己的存档)")
                log.errorWithTraceBack(info)
            end
            if ____exports.default.verifyPlayerName and data._SH and data._SH ~= shTemp then
                if data._SI and data._SI > 0 and data._SI == userId then
                    print("检测到玩家改名了！自动修正存档!")
                else
                    data = {}
                    DisplayTimedTextToPlayer(
                        tempPlayer,
                        0,
                        0,
                        60,
                        "存档验证失败!自动退出游戏!"
                    )
                    BaseUtil.runLater(
                        1,
                        function()
                            CustomDefeatBJ(tempPlayer, "警告:存档验证失败C2！")
                        end
                    )
                end
            end
            return true, data
        end)
        if not ____try then
            ____hasReturned, ____returnValue = ____catch(____hasReturned)
        end
        if ____hasReturned then
            return ____returnValue
        end
    end
    return data
end
ArchiveUtil.encrypt = true
ArchiveUtil.verifyPlayerName = not isDebug
ArchiveUtil.ArchiveKeyCountMax = 150
ArchiveUtil.salt = "s_z_b_s_q_j"
ArchiveUtil.baseArchiveKeyPrefix = "Solar_Base_Archive_"
ArchiveUtil.playerDatas = {}
ArchiveUtil.isInit = false
____exports.default = ArchiveUtil
return ____exports

File: src/solar/util/archive/ArchiveUtil.ts
/** @noSelf **/
import CodecUtil from "../math/CodecUtil";
import BaseUtil from "../BaseUtil";
import {MapPlayer} from "solar/w3ts/handles/player";

/**
 将存档key 与存档值 作为Json字符串 统一存储到平台服务器上
 因此请尽量减少key的长度 以多存内容

 修改玩家名之后将导致存档失效

 已被占用的key: _SL  _SH _SI//用作存储存档玩家的地图等级 玩家名字哈希 和玩家账号id 防止盒子佬一键复制别人的存档
 */
export default class ArchiveUtil {
    //加密 请在地图发布前就确定是否需要加密以及加密的方式。发布后不能更改此选项 否则会造成旧存档被清空
    static encrypt: boolean = true;
    static verifyPlayerName: boolean = !isDebug;
    static ArchiveKeyCountMax: number = 150;
    static salt: string = "s_z_b_s_q_j";
    static baseArchiveKeyPrefix: string = "Solar_Base_Archive_";
    //
    static playerDatas: { [playerIndex: string]: { [key: string]: any } } = {}


    /**
     * 获取玩家地图存档 没有则返回null
     * @param whichPlayer
     * @param key
     * @param mapLevelLimit 地图等级限制（当玩家地图等级小于此数字时返回null）
     */
    static get(whichPlayer: player, key: string, mapLevelLimit: number = 0) {
        ArchiveUtil._init0();
        if (mapLevelLimit > 0 && DzAPI_Map_GetMapLevel(whichPlayer) < mapLevelLimit) {
            return null;
        }
        let data = ArchiveUtil.playerDatas["P" + GetPlayerId(whichPlayer)];
        if (!data) {
            return null;
        }
        return data[key];
    }

    /**
     * 保存玩家的一个存档
     * @param whichPlayer
     * @param key
     * @param value
     * @param isSaveAll 是否立即上传到服务器存档(批量设置值时可以先填false 最后再统一上传 节省一丝性能)
     */
    static set(whichPlayer: player, key: string, value: any, isSaveAll: boolean = true) {
        ArchiveUtil._init0();
        let data = ArchiveUtil.playerDatas["P" + GetPlayerId(whichPlayer)];
        if (!data) {
            data = {}
            ArchiveUtil.playerDatas["P" + GetPlayerId(whichPlayer)] = data;
        }
        data[key] = value;
        if (isSaveAll) {
            ArchiveUtil.saveAll(whichPlayer)
        }
    }

    /**
     *获取所有存档
     */
    static getAllArchive(whichPlayer: player): any {
        ArchiveUtil._init0();
        return ArchiveUtil.playerDatas["P" + GetPlayerId(whichPlayer)];
    }

    /**
     * 打印某个玩家的所有存档 调试时用
     * @param whichPlayer
     */
    static printAllArchive(whichPlayer: player): void {
        ArchiveUtil._init0();
        let data = ArchiveUtil.playerDatas["P" + GetPlayerId(whichPlayer)];
        if (!data) {
            return;
        }
        print("========打印玩家" + (GetPlayerId(whichPlayer) + 1) + "的存档 开始========")
        for (let dataKey in data) {
            print(dataKey + "=[" + data[dataKey] + "]")
        }
        print("========打印玩家" + (GetPlayerId(whichPlayer) + 1) + "的存档 结束========")
    }

    /**
     * 清空所有存档
     */
    static clearAllArchive(whichPlayer: player): void {
        for (let i = 0; i < ArchiveUtil.ArchiveKeyCountMax; i++) {
            let key: string = ArchiveUtil.baseArchiveKeyPrefix + i;
            DzAPI_Map_FlushStoredMission(whichPlayer, key);
        }
        print("已清理全部存档!")
    }

    /**
     * 保存玩家的所有存档数据
     */
    static saveAll(whichPlayer: player): void {
        let data = ArchiveUtil.playerDatas["P" + GetPlayerId(whichPlayer)];
        if (!data) {
            data = {}
            ArchiveUtil.playerDatas["P" + GetPlayerId(whichPlayer)] = data;
        }
        //设置地图等级 以作存档验证  _SI
        data._SL = DzAPI_Map_GetMapLevel(whichPlayer);
        data._SH = StringHash(GetPlayerName(whichPlayer));
        data._SI = MapPlayer.fromHandle(whichPlayer).userId;

        let dataJsonStr: string = JSON.stringify(data);
        if (ArchiveUtil.encrypt) {
            dataJsonStr = CodecUtil.saesEncode(dataJsonStr, gameName + ArchiveUtil.salt)
        }
        let dataJsonStrLength = dataJsonStr.length;
        let keyLength = math.floor(dataJsonStrLength / 64) + 1;
        log.debug("玩家" + GetPlayerId(whichPlayer) + "的存档栏位大小为:" + keyLength)
        if (keyLength > ArchiveUtil.ArchiveKeyCountMax) {
            log.errorWithTraceBack("存档栏位大小超过" + ArchiveUtil.ArchiveKeyCountMax + "！" + keyLength)
            return;
        }
        for (let i = 0; i < keyLength; i++) {
            let endIndex = (i + 1) * 64;
            if (endIndex > dataJsonStrLength) {
                endIndex = dataJsonStrLength;
            }
            let oneData = dataJsonStr.substring(i * 64, endIndex)
            let key: string = ArchiveUtil.baseArchiveKeyPrefix + i;
            DzAPI_Map_SaveServerValue(whichPlayer, key, oneData);
        }
        //清空后一个存档栏位 以标记结束
        let key: string = ArchiveUtil.baseArchiveKeyPrefix + keyLength;
        DzAPI_Map_FlushStoredMission(whichPlayer, key);
    }


    /**
     * 初始化存档数据
     * @private
     */
    private static isInit = false;

    private static _init0(): boolean {
        if (ArchiveUtil.isInit) {
            return true;
        }
        ArchiveUtil.isInit = true;
        for (let i = 0; i < bj_MAX_PLAYER_SLOTS; i++) {
            let tempPlayer = Player(i);

            if (GetPlayerController(tempPlayer) == MAP_CONTROL_USER && GetPlayerSlotState(tempPlayer) == PLAYER_SLOT_STATE_PLAYING) {
                ArchiveUtil.playerDatas["P" + i] = ArchiveUtil._init0_onePlayer(tempPlayer);

            }
        }
        return true;
    }

    private static _init0_onePlayer(tempPlayer: player) {
        let data: any = {};
        let i = GetPlayerId(tempPlayer);
        let dataJsonStr: string = "";
        for (let i = 0; i < ArchiveUtil.ArchiveKeyCountMax; i++) {
            let key: string = ArchiveUtil.baseArchiveKeyPrefix + i;
            let serverValue: string = DzAPI_Map_GetServerValue(tempPlayer, key);
            if (serverValue && serverValue.length > 0) {
                dataJsonStr = dataJsonStr + serverValue;
            } else {
                break;
            }
        }
        //转json了
        if (dataJsonStr.length < 2) {
            return {};
        }
        try {
            if (ArchiveUtil.encrypt) {
                dataJsonStr = CodecUtil.saesDecode(dataJsonStr, gameName + ArchiveUtil.salt)
            }
            data = JSON.parse(dataJsonStr);
            log.debug("玩家" + (i + 1) + ":的存档数据为:" + dataJsonStr)
            //验证存档
            if (DzAPI_Map_GetMapLevel(tempPlayer) < data._SL) {
                print("存档验证失败!默认使用空存档!")
                data = {}
                BaseUtil.runLater(1, () => {
                    CustomDefeatBJ(tempPlayer, "警告:存档验证失败C1！");
                })
            }
            //验证玩家名哈希
            let shTemp = StringHash(GetPlayerName(tempPlayer));
            let userId = MapPlayer.fromHandle(tempPlayer).userId;
            if (userId == -1) {
                //userId 还在获取同步中...
                const info = "提示:请不要再地图初始时获取存档！应当延迟执行存档逻辑!";
                // DisplayTimedTextToPlayer(GetLocalPlayer(), 0, 0, 10, info);
                print(info)
                print("请等待内置userId(用户id)同步完成！(为了解决玩家名更改后靠userId判断是否是自己的存档)")
                log.errorWithTraceBack(info)
            }
            //校验存档
            if (ArchiveUtil.verifyPlayerName && data._SH && data._SH != shTemp) {//如果玩家名哈希不一样  则说明玩家改名或则复制了别人的存档
                if (data._SI && data._SI > 0 && data._SI == userId) {//说明存档就是本玩家的
                    print("检测到玩家改名了！自动修正存档!")
                } else {
                    //存档有问题的玩家 自己清空存档并退出游戏
                    data = {}
                    DisplayTimedTextToPlayer(tempPlayer, 0, 0, 60, '存档验证失败!自动退出游戏!');
                    BaseUtil.runLater(1, () => {
                        CustomDefeatBJ(tempPlayer, "警告:存档验证失败C2！");
                    })
                }
            }
            //设置
            return data;
        } catch (e) {
            log.errorWithTraceBack("存档Json格式解析出错!")
            DisplayTimedTextToPlayer(GetLocalPlayer(), 0, 0, 60, '存档格式解析出错！');
            print("需要解析的Json字符串为[" + dataJsonStr + "]")
        }
        return data;
    }
}
